File: Sources/App/Classes/Dialogs/Channel Spotlight/TDCChannelSpotlightSearchResultsTable.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2010 - 2020 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import "TXGlobalModels.h"
#import "NSViewHelper.h"
#import "TLOLocalization.h"
#import "IRCClient.h"
#import "IRCChannel.h"
#import "TDCChannelSpotlightAppearancePrivate.h"
#import "TDCChannelSpotlightControllerInternal.h"
#import "TDCChannelSpotlightSearchResultPrivate.h"
#import "TDCChannelSpotlightSearchResultsTablePrivate.h"

NS_ASSUME_NONNULL_BEGIN

@interface TDCChannelSpotlightSearchResultRowView ()
@property(nonatomic, weak) TDCChannelSpotlightController *controller;
@property(nonatomic, weak) TDCChannelSpotlightSearchResultCellView *childCell;
@property(readonly) TDCChannelSpotlightAppearance *userInterfaceObjects;
@end

@interface TDCChannelSpotlightSearchResultCellView ()
@property(readonly, copy) NSAttributedString *channelName;
@property(readonly, copy) NSString *keyboardShortcut;
@property(readonly, copy) NSString *unreadCountDescription;
@property(nonatomic, weak) IBOutlet NSTextField *channelNameField;
@property(nonatomic, weak) IBOutlet NSTextField *keyboardShortcutField;
@property(nonatomic, weak) IBOutlet NSLayoutConstraint *keyboardShortcutFieldOffsetConstraint;
@property(nonatomic, weak) IBOutlet NSTextField *unreadCountDescriptionField;
@property(nonatomic, assign) BOOL staticLabelsPopulated;
@property(nonatomic, strong, nullable) IRCChannel *observedChannel;
@property(readonly) TDCChannelSpotlightAppearance *userInterfaceObjects;
@property(readonly) TDCChannelSpotlightController *controller;
@property(readonly) TDCChannelSpotlightSearchResultRowView *rowCell;
@end

@implementation TDCChannelSpotlightSearchResultCellView

- (BOOL)wantsLayer
{
	return YES;
}

- (NSViewLayerContentsRedrawPolicy)layerContentsRedrawPolicy
{
	return NSViewLayerContentsRedrawOnSetNeedsDisplay;
}

- (void)setInitialValues
{
	[self channelNameChanged];
}

- (void)updateLayer
{
	[self updateAppearance];

	[self keyboardShortcutChanged];
}

- (void)updateAppearance
{
	TDCChannelSpotlightSearchResultRowView *rowCell = self.rowCell;

	[self updateTextFieldTextColorIsSelected:rowCell.isSelected];
}

- (void)updateTextFieldTextColorIsSelected:(BOOL)isSelected
{
	TDCChannelSpotlightAppearance *appearance = self.userInterfaceObjects;

	TDCChannelSpotlightSearchResultRowView *rowCell = self.rowCell;

	/* Selection is drawn by AppKit. Only the text colours follow it:
	 an emphasized (key window) selection is drawn with the accent
	 colour and needs the alternate text colour, every other state
	 uses the vibrant label colours. */
	if (isSelected && rowCell.isEmphasized) {
		NSColor *selectedTextColor = [NSColor alternateSelectedControlTextColor];

		self.channelNameField.textColor = selectedTextColor;
		self.unreadCountDescriptionField.textColor = selectedTextColor;
		self.keyboardShortcutField.textColor = selectedTextColor;
	} else {
		self.channelNameField.textColor = [NSColor labelColor];
		self.unreadCountDescriptionField.textColor = [NSColor secondaryLabelColor];
		self.keyboardShortcutField.textColor = (isSelected) ? [NSColor labelColor] : [NSColor secondaryLabelColor];
	}

	if (isSelected) {
		self.keyboardShortcutFieldOffsetConstraint.constant = appearance.searchResultKeyboardShortcutSelectedOffset;
	} else {
		self.keyboardShortcutFieldOffsetConstraint.constant = appearance.searchResultKeyboardShortcutDeselectedOffset;
	}
}

- (NSAttributedString *)channelName
{
	TDCChannelSpotlightSearchResult *searchResult = self.objectValue;

	if (searchResult == nil) {
		return [NSAttributedString attributedString];
	}

	static NSMutableParagraphStyle *paragraphStyle = nil;

	if (paragraphStyle == nil) {
		paragraphStyle = [[NSParagraphStyle defaultParagraphStyle] mutableCopy];

		paragraphStyle.lineBreakMode = NSLineBreakByTruncatingTail;
	}

	NSString *channelName = searchResult.channel.name;

	NSFont *channelNameFieldFont = self.channelNameField.font;

	NSMutableAttributedString *resultString = [NSMutableAttributedString
		mutableAttributedStringWithString:TXTLS(@"TDCChannelSpotlightController[jpw-cj]", channelName)
							   attributes:@{
								   NSFontAttributeName : channelNameFieldFont,
								   NSParagraphStyleAttributeName : paragraphStyle
							   }];

	TDCChannelSpotlightController *controller = self.controller;

	NSString *searchString = controller.searchString;

	[resultString.string enumerateFirstOccurrenceOfCharactersInString:searchString
															withBlock:^(NSRange range, BOOL *stop) {
																NSFont *boldFont =
																	[RZFontManager() convertFont:channelNameFieldFont
																					 toHaveTrait:NSBoldFontMask];

																[resultString addAttribute:NSFontAttributeName
																					 value:boldFont
																					 range:range];
															}
															  options:NSCaseInsensitiveSearch];

	NSString *networkName = searchResult.channel.associatedClient.networkNameAlt;

	[resultString appendString:TXTLS(@"TDCChannelSpotlightController[z68-5q]", networkName)];

	return resultString;
}

- (void)channelNameChanged
{
	[self willChangeValueForKey:@"channelName"];
	[self didChangeValueForKey:@"channelName"];
}

- (NSString *)keyboardShortcut
{
	TDCChannelSpotlightSearchResult *searchResult = self.objectValue;

	if (searchResult == nil) {
		return @"";
	}

	TDCChannelSpotlightController *controller = self.controller;

	NSArray *searchResults = controller.searchResultsFiltered;

	NSUInteger searchResultIndex = [searchResults indexOfObjectIdenticalTo:searchResult];

	if (searchResultIndex != NSNotFound && (NSInteger)searchResultIndex == controller.selectedSearchResult) {
		return @"↩︎";
	}

	if (searchResultIndex > 9) {
		return @"";
	}

	NSUInteger keyboardShortcutIndex = (searchResultIndex + 1);

	if (keyboardShortcutIndex == 10) {
		keyboardShortcutIndex = 0;
	}

	return [NSString stringWithFormat:@"⌘%lu", keyboardShortcutIndex];
}

- (void)keyboardShortcutChanged
{
	[self willChangeValueForKey:@"keyboardShortcut"];
	[self didChangeValueForKey:@"keyboardShortcut"];
}

- (NSString *)unreadCountDescription
{
	TDCChannelSpotlightSearchResult *searchResult = self.objectValue;

	if (searchResult == nil) {
		return @"";
	}

	NSUInteger nicknameHighlightCount = searchResult.channel.nicknameHighlightCount;

	NSString *nicknameHighlightCountDescription = nil;

	if (nicknameHighlightCount == 1) {
		nicknameHighlightCountDescription =
			TXTLS(@"TDCChannelSpotlightController[0lz-oh]", TXFormattedNumber(nicknameHighlightCount));
	} else {
		nicknameHighlightCountDescription =
			TXTLS(@"TDCChannelSpotlightController[c4u-21]", TXFormattedNumber(nicknameHighlightCount));
	}

	NSUInteger unreadCount = searchResult.channel.treeUnreadCount;

	NSString *unreadCountDescription = nil;

	if (unreadCount == 1) {
		unreadCountDescription = TXTLS(@"TDCChannelSpotlightController[43s-x4]", TXFormattedNumber(unreadCount));
	} else {
		unreadCountDescription = TXTLS(@"TDCChannelSpotlightController[vzj-30]", TXFormattedNumber(unreadCount));
	}

	return TXTLS(@"TDCChannelSpotlightController[et7-c5]", nicknameHighlightCountDescription, unreadCountDescription);
}

- (void)unreadCountDescriptionChanged
{
	[self willChangeValueForKey:@"unreadCountDescription"];
	[self didChangeValueForKey:@"unreadCountDescription"];

	[self updateAccessibilityLabel];
}

- (void)updateAccessibilityLabel
{
	TDCChannelSpotlightSearchResult *searchResult = self.objectValue;

	if (searchResult == nil) {
		self.accessibilityLabel = nil;

		return;
	}

	NSString *channelName = searchResult.channel.name;

	NSString *networkName = searchResult.channel.associatedClient.networkNameAlt;

	NSString *label = [TXTLS(@"TDCChannelSpotlightController[jpw-cj]", channelName)
		stringByAppendingString:TXTLS(@"TDCChannelSpotlightController[z68-5q]", networkName)];

	self.accessibilityLabel = TXTLS(@"TDCChannelSpotlightController[et7-c5]", label, self.unreadCountDescription);
}

- (TDCChannelSpotlightSearchResultRowView *)rowCell
{
	return (id)self.superview;
}

- (TDCChannelSpotlightAppearance *)userInterfaceObjects
{
	return self.rowCell.userInterfaceObjects;
}

- (TDCChannelSpotlightController *)controller
{
	return self.rowCell.controller;
}

- (void)viewDidMoveToWindow
{
	[super viewDidMoveToWindow];

	[self updateObservedChannel];
}

- (void)setObjectValue:(nullable id)objectValue
{
	[super setObjectValue:objectValue];

	/* Cells are reused. The channel this cell represents can change
	 after the cell has already been placed in a window, so the
	 observation must follow the object value, not the window. */
	[self updateObservedChannel];
}

- (void)prepareForReuse
{
	[super prepareForReuse];

	[self stopObservingChannel];
}

- (void)dealloc
{
	[self stopObservingChannel];
}

- (void)updateObservedChannel
{
	IRCChannel *channel = nil;

	if (self.window != nil) {
		TDCChannelSpotlightSearchResult *searchResult = self.objectValue;

		channel = searchResult.channel;
	}

	if (channel == self.observedChannel) {
		return;
	}

	[self stopObservingChannel];

	if (channel == nil) {
		return;
	}

	self.observedChannel = channel;

	[channel addObserver:self
			  forKeyPath:@"nicknameHighlightCount"
				 options:(NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew)
				 context:nil];
	[channel addObserver:self
			  forKeyPath:@"treeUnreadCount"
				 options:(NSKeyValueObservingOptionInitial | NSKeyValueObservingOptionNew)
				 context:nil];

	[self setInitialValues];
}

- (void)stopObservingChannel
{
	IRCChannel *channel = self.observedChannel;

	if (channel == nil) {
		return;
	}

	[channel removeObserver:self forKeyPath:@"nicknameHighlightCount"];
	[channel removeObserver:self forKeyPath:@"treeUnreadCount"];

	self.observedChannel = nil;
}

- (void)observeValueForKeyPath:(nullable NSString *)keyPath
					  ofObject:(nullable id)object
						change:(nullable NSDictionary<NSString *, id> *)change
					   context:(nullable void *)context
{
	if ([keyPath isEqualToString:@"nicknameHighlightCount"] || [keyPath isEqualToString:@"treeUnreadCount"]) {
		[self unreadCountDescriptionChanged];
	}
}

@end

#pragma mark -

@implementation TDCChannelSpotlightSearchResultRowView

- (instancetype)initWithController:(TDCChannelSpotlightController *)controller
{
	NSParameterAssert(controller != nil);

	if ((self = [super initWithFrame:NSZeroRect])) {
		self.controller = controller;

		return self;
	}

	return nil;
}

- (void)setSelected:(BOOL)selected
{
	super.selected = selected;

	if (selected == NO && self.invalidatingBackgroundForSelection) {
		return;
	}

	[self setNeedsDisplayOnChild];
}

- (void)setNeedsDisplayOnChild
{
	self.childCell.needsDisplay = YES;
}

- (void)setEmphasized:(BOOL)emphasized
{
	super.emphasized = emphasized;

	[self setNeedsDisplayOnChild];
}

- (nullable TDCChannelSpotlightSearchResultCellView *)childCell
{
	if (self->_childCell == nil) {
		if (self.numberOfColumns == 0) {
			return nil;
		}

		self->_childCell = [self viewAtColumn:0];
	}

	return self->_childCell;
}

- (TDCChannelSpotlightAppearance *)userInterfaceObjects
{
	return self.controller.userInterfaceObjects;
}

@end

NS_ASSUME_NONNULL_END

File: Sources/App/Modules/AppKitSupport/TDCChannelSpotlightSearchResultsTable.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \\ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\\___/_/\\_\\__|\\__,_|\\__,_|_|
 *
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 *********************************************************************** */

import AppKit
import CocoaExtensions

@objc(TDCChannelSpotlightSearchResultCellView)
public final class ChannelSpotlightSearchResultCellView: NSTableCellView {
	@IBOutlet private var channelNameField: NSTextField!
	@IBOutlet private var keyboardShortcutField: NSTextField!
	@IBOutlet private var keyboardShortcutFieldOffsetConstraint: NSLayoutConstraint!
	@IBOutlet private var unreadCountDescriptionField: NSTextField!

	private var observedChannel: IRCChannel?
	private var channelObservations: [NSKeyValueObservation] = []

	override public var wantsLayer: Bool {
		get { true }
		set {}
	}

	override public var layerContentsRedrawPolicy: NSView.LayerContentsRedrawPolicy {
		get { .onSetNeedsDisplay }
		set {}
	}

	override public func updateLayer() {
		updateAppearance()
		keyboardShortcutChanged()
	}

	private func setInitialValues() {
		channelNameChanged()
	}

	private func updateAppearance() {
		updateTextFieldTextColor(isSelected: rowCell.isSelected)
	}

	private func updateTextFieldTextColor(isSelected: Bool) {
		let appearance = userInterfaceObjects
		let rowCell = rowCell

		/* Selection is drawn by AppKit. Only the text colours follow it:
		 an emphasized (key window) selection is drawn with the accent
		 colour and needs the alternate text colour, every other state
		 uses the vibrant label colours. */
		if isSelected, rowCell.isEmphasized {
			let selectedTextColor = NSColor.alternateSelectedControlTextColor
			channelNameField.textColor = selectedTextColor
			unreadCountDescriptionField.textColor = selectedTextColor
			keyboardShortcutField.textColor = selectedTextColor
		} else {
			channelNameField.textColor = .labelColor
			unreadCountDescriptionField.textColor = .secondaryLabelColor
			keyboardShortcutField.textColor = isSelected ? .labelColor : .secondaryLabelColor
		}

		if isSelected {
			keyboardShortcutFieldOffsetConstraint.constant = appearance.searchResultKeyboardShortcutSelectedOffset
		} else {
			keyboardShortcutFieldOffsetConstraint.constant = appearance.searchResultKeyboardShortcutDeselectedOffset
		}
	}

	@objc public var channelName: NSAttributedString {
		guard let searchResult = objectValue as? ChannelSpotlightSearchResult,
		      let channel = searchResult.channel
		else {
			return NSAttributedString()
		}

		guard let paragraphStyle = NSParagraphStyle.default.mutableCopy() as? NSMutableParagraphStyle else {
			return NSAttributedString()
		}
		paragraphStyle.lineBreakMode = .byTruncatingTail

		let channelName = channel.name
		let channelNameFieldFont = channelNameField.font ?? NSFont.systemFont(ofSize: NSFont.systemFontSize)

		let resultString = NSMutableAttributedString(
			string: ChannelSpotlightStrings.channelName(channelName),
			attributes: [
				.font: channelNameFieldFont,
				.paragraphStyle: paragraphStyle,
			]
		)

		let searchString = controller.searchString
		let matchedRanges = resultString.string.rangesOfFirstOccurrences(
			ofCharactersIn: searchString,
			options: .caseInsensitive
		)

		if !matchedRanges.isEmpty {
			let boldFont = NSFontManager.shared.convert(channelNameFieldFont, toHaveTrait: .boldFontMask)

			for range in matchedRanges {
				resultString.addAttribute(.font, value: boldFont, range: range)
			}
		}

		let networkName = channel.associatedClient?.networkNameAlt ?? ""
		resultString.append(
			NSAttributedString(string: ChannelSpotlightStrings.networkSuffix(networkName))
		)

		return resultString
	}

	private func channelNameChanged() {
		willChangeValue(forKey: "channelName")
		didChangeValue(forKey: "channelName")
	}

	@objc public var keyboardShortcut: String {
		guard let searchResult = objectValue as? ChannelSpotlightSearchResult else {
			return ""
		}

		let searchResults = controller.searchResultsFiltered
		guard let searchResultIndex = searchResults.firstIndex(where: { $0 === searchResult }) else {
			return ""
		}

		if searchResultIndex == controller.selectedSearchResult {
			return "↩︎"
		}

		if searchResultIndex > 9 {
			return ""
		}

		var keyboardShortcutIndex = searchResultIndex + 1
		if keyboardShortcutIndex == 10 {
			keyboardShortcutIndex = 0
		}

		return "⌘\(keyboardShortcutIndex)"
	}

	private func keyboardShortcutChanged() {
		willChangeValue(forKey: "keyboardShortcut")
		didChangeValue(forKey: "keyboardShortcut")
	}

	@objc public var unreadCountDescription: String {
		guard let searchResult = objectValue as? ChannelSpotlightSearchResult,
		      let channel = searchResult.channel
		else {
			return ""
		}

		let nicknameHighlightCount = channel.nicknameHighlightCount
		let nicknameHighlightCountDescription = ChannelSpotlightStrings.highlights(Int(nicknameHighlightCount))

		let unreadCount = channel.treeUnreadCount
		let unreadCountDescription = ChannelSpotlightStrings.unreadMessages(Int(unreadCount))

		return ChannelSpotlightStrings.combined(
			nicknameHighlightCountDescription,
			unreadCountDescription
		)
	}

	private func unreadCountDescriptionChanged() {
		willChangeValue(forKey: "unreadCountDescription")
		didChangeValue(forKey: "unreadCountDescription")
		updateAccessibilityLabel()
	}

	private func updateAccessibilityLabel() {
		guard let searchResult = objectValue as? ChannelSpotlightSearchResult,
		      let channel = searchResult.channel
		else {
			setAccessibilityLabel(nil)
			return
		}

		let channelName = channel.name
		let networkName = channel.associatedClient?.networkNameAlt ?? ""
		let label =
			ChannelSpotlightStrings.channelName(channelName)
				+ ChannelSpotlightStrings.networkSuffix(networkName)

		setAccessibilityLabel(ChannelSpotlightStrings.combined(label, unreadCountDescription))
	}

	private var rowCell: ChannelSpotlightSearchResultRowView {
		guard let rowCell = superview as? ChannelSpotlightSearchResultRowView else {
			preconditionFailure("Search-result content must be hosted by its row view")
		}

		return rowCell
	}

	private var userInterfaceObjects: ChannelSpotlightAppearance {
		rowCell.userInterfaceObjects
	}

	private var controller: ChannelSpotlightController {
		rowCell.controller
	}

	override public func viewDidMoveToWindow() {
		super.viewDidMoveToWindow()
		updateObservedChannel()
	}

	override public var objectValue: Any? {
		didSet {
			/* Cells are reused. The channel this cell represents can change
			 after the cell has already been placed in a window, so the
			 observation must follow the object value, not the window. */
			updateObservedChannel()
		}
	}

	override public func prepareForReuse() {
		super.prepareForReuse()
		stopObservingChannel()
	}

	private func updateObservedChannel() {
		var channel: IRCChannel?

		if window != nil, let searchResult = objectValue as? ChannelSpotlightSearchResult {
			channel = searchResult.channel
		}

		if channel === observedChannel {
			return
		}

		stopObservingChannel()

		guard let channel else {
			return
		}

		observedChannel = channel
		channelObservations = [
			channel.observe(\.nicknameHighlightCount, options: [.initial, .new]) { [weak self] _, _ in
				/* ISOLATION-EXCEPTION: `NSKeyValueObservation`'s change handler is
				 nonisolated. The observed counts are only mutated on the main actor. */
				MainActor.assumeIsolated {
					self?.unreadCountDescriptionChanged()
				}
			},
			channel.observe(\.treeUnreadCount, options: [.initial, .new]) { [weak self] _, _ in
				/* ISOLATION-EXCEPTION: `NSKeyValueObservation`'s change handler is
				 nonisolated. The observed counts are only mutated on the main actor. */
				MainActor.assumeIsolated {
					self?.unreadCountDescriptionChanged()
				}
			},
		]

		setInitialValues()
	}

	private func stopObservingChannel() {
		channelObservations.forEach { $0.invalidate() }
		channelObservations.removeAll()
		observedChannel = nil
	}
}

// MARK: -

@objc(TDCChannelSpotlightSearchResultRowView)
public final class ChannelSpotlightSearchResultRowView: NSTableRowView {
	@objc public private(set) weak var controller: ChannelSpotlightController!
	private weak var childCellStorage: ChannelSpotlightSearchResultCellView?

	@objc(initWithController:)
	public init(controller: ChannelSpotlightController) {
		self.controller = controller
		super.init(frame: .zero)
	}

	@available(*, unavailable)
	public required init?(coder _: NSCoder) {
		fatalError("init(coder:) has not been implemented")
	}

	override public var isSelected: Bool {
		didSet {
			setNeedsDisplayOnChild()
		}
	}

	private func setNeedsDisplayOnChild() {
		childCell?.needsDisplay = true
	}

	override public var isEmphasized: Bool {
		didSet {
			setNeedsDisplayOnChild()
		}
	}

	private var childCell: ChannelSpotlightSearchResultCellView? {
		if childCellStorage == nil {
			if numberOfColumns == 0 {
				return nil
			}

			childCellStorage = view(atColumn: 0) as? ChannelSpotlightSearchResultCellView
		}

		return childCellStorage
	}

	@objc public var userInterfaceObjects: ChannelSpotlightAppearance {
		controller.userInterfaceObjects
	}
}
